Correct .gitattributes

JunielKatarn opened the pull request:
Fixes #
Upon clone, new git attributes file messes up the modified state of sources, marking dozens of the as modified on checkout, and making the local clone unable to reset to an unmodified state.

See sample freshly checked-out master clone:
![image](https://user-images.githubusercontent.com/4507319/110873608-f99d5b80-8286-11eb-9d2b-ea36275e8f56.png)

With this change, checked-out clones are clean from the beginning.
It may be worth investigate what may cause this and re-introduce a fixed `.gitattributes` later on.

## Proposed Changes

  - Set `text=auto` only on the affected extensions (the ones that need CRLF enforcement).
  
  @vgturtle127 can you take a look?

flagrama commented:
Removing `* text=auto` is probably correct.  I'm not sure it's necessary to remove the entire file though.

I forgot that `* text=auto` forces working-copy to be `LF` but without that, it will accept `CR LF` in working tree. It's useful to have binary files explicitly considered binary and text-based file text to prevent issues where Git decides a file looks suspiciously binary or text when it's actually the opposite. This could cause a binary to be modified with any bytes that look like ASCII "CR LF" to be converted to just ASCII "LF" which could break a binary file. If a text file is seen as binary it just won't be converted correctly, which could be annoying while trying to keep the Linux project64-core buildable until it gets to the point where a cross-platform GUI starts being implemented.

vgturtle127 commented:
Yeah, I guess removing * text=auto is the best option. I had thought that it would work seamlessly but I guess I was wrong. I always thought it converted the line endings on the users machine, and then converted them back when commiting.

So I think removing * text=auto is the best option since there are definitely things we don't want to be normalized by Git.

JunielKatarn commented:
> Yeah, I guess removing * text=auto is the best option. I had thought that it would work seamlessly but I guess I was wrong. I always thought it converted the line endings on the users machine, and then converted them back when commiting.
> 
> So I think removing * text=auto is the best option since there are definitely things we don't want to be normalized by Git.

Removing `* text=auto` didn't do the trick. I get the same behavior.
Most projects I know don't have a `.gitattributes` file. Maybe we can manage.

JunielKatarn commented:
> Removing `* text=auto` is probably correct. I'm not sure it's necessary to remove the entire file though.
> 
> I forgot that `* text=auto` forces working-copy to be `LF` but without that, it will accept `CR LF` in working tree. It's useful to have binary files explicitly considered binary and text-based file text to prevent issues where Git decides a file looks suspiciously binary or text when it's actually the opposite. This could cause a binary to be modified with any bytes that look like ASCII "CR LF" to be converted to just ASCII "LF" which could break a binary file. If a text file is seen as binary it just won't be converted correctly, which could be annoying while trying to keep the Linux project64-core buildable until it gets to the point where a cross-platform GUI starts being implemented.

For that purpose, setting `core.autocrlf` to `false` should do the trick.

vgturtle127 commented:
I would say 99% of projects on GitHub and that use Git in general either already do or should have a .gitattributes file.

I can't replicate the issue you are having though, so I'm not sure what is causing that on your end 😕

vgturtle127 commented:
IDK, it's weird that removing the change did nothing, in that case does everything else work for you @JunielKatarn ?

JunielKatarn commented:
> I can't replicate the issue you are having though, so I'm not sure what is causing that on your end 😕

```
git clone git@github.com:project64/project64.git
cd .\project64\
```
![image](https://user-images.githubusercontent.com/4507319/110892815-9bce3b00-82a9-11eb-8af4-b5abe0ba4128.png)

```
 git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   Source/Project64-video/Renderer/OGLcombiner.cpp
        modified:   Source/Project64-video/Renderer/OGLgeometry.cpp
        modified:   Source/Project64-video/Renderer/OGLglitchmain.cpp
        modified:   Source/Project64-video/Renderer/OGLtextures.cpp
        modified:   Source/Project64-video/Renderer/glitchmain.h
        modified:   Source/Project64-video/TexBuffer.cpp
        modified:   Source/Project64-video/TexBuffer.h
        modified:   Source/Project64-video/TexCache.cpp
        modified:   Source/Project64-video/TexCache.h
        modified:   Source/Project64-video/TexConv.h
        modified:   Source/Project64-video/TexLoad.h
        modified:   Source/Project64-video/TexLoad16b.h
        modified:   Source/Project64-video/TexLoad32b.h
        modified:   Source/Project64-video/TexLoad4b.h
        modified:   Source/Project64-video/TexLoad8b.h
        modified:   Source/Project64-video/TexMod.h
        modified:   Source/Project64-video/TexModCI.h
        modified:   Source/Project64-video/TextureEnhancer/TxUtil.cpp
        modified:   Source/Project64-video/Util.cpp
        modified:   Source/Project64-video/Util.h
        modified:   Source/Project64-video/rdp.cpp
        modified:   Source/Project64-video/rdp.h
        modified:   Source/Project64-video/resource.h
        modified:   Source/Project64-video/turbo3D.cpp
        modified:   Source/Project64-video/ucode.cpp
        modified:   Source/Project64-video/ucode00.cpp
        modified:   Source/Project64-video/ucode01.cpp
        modified:   Source/Project64-video/ucode02.cpp
        modified:   Source/Project64-video/ucode03.cpp
        modified:   Source/Project64-video/ucode04.cpp
        modified:   Source/Project64-video/ucode05.cpp
        modified:   Source/Project64-video/ucode06.cpp
        modified:   Source/Project64-video/ucode07.cpp
        modified:   Source/Project64-video/ucode08.cpp
        modified:   Source/Project64-video/ucode09.cpp
        modified:   Source/Project64-video/ucode09rdp.cpp
        modified:   Source/Project64-video/ucodeFB.cpp
        modified:   Source/Project64/UserInterface/Debugger/Breakpoints.cpp
        modified:   Source/Project64/UserInterface/Settings/SettingsPage-Game-Plugin.cpp
        modified:   Source/Project64/UserInterface/Settings/SettingsPage-Game-Plugin.h
        modified:   Source/Project64/UserInterface/WTLControls/DisplayMode.h
```

I tried:
```
git checkout .
git reset --hard
* remove the text=auto line  from .gitattributes and commit *
```
... same result.

Finally:
```
* git rm .gitattributes *
```
Issue is resolved.

I'll try one more thing.

JunielKatarn commented:
> IDK, it's weird that removing the change did nothing, in that case does everything else work for you @JunielKatarn ?

Can you please clone the latest master and try `git status`?

Also, I use the normal Windows environment (Windows Powershell).
```
git --version
git version 2.28.0.windows.1

git log -n 1
commit 33011f3a572a551a089a09a422f5c43ace02a084
Merge: 666b2c64 64c3bcfc
Author: zilmar <zilmar@pj64-emu.com>
Date:   Fri Mar 12 07:18:29 2021 +1030

    Merge pull request #2005 from vgturtle127/master

    General project improvements and updates
```

What is your dev environment, @vgturtle127?

vgturtle127 commented:
I'm using Git Bash and it says nothing to commit, working tree clean.

I'm on 2.30.2.windows.1

flagrama commented:
After committing the removal of the * text I can clone the repository fine. You're reusing an already contaminated repository in your test. I think it might store something in the .git directory when cloning with that line in the gitattributes which continues to normalize line endings.

vgturtle127 commented:
I still had the * text=auto in my repo when I cloned it, so not sure if this is a bug or if I'm not doing something I should be or what is happening heh

JunielKatarn commented:
> After committing the removal of the * text I can clone the repository fine. You're reusing an already contaminated repository in your test. I think it might store something in the .git directory when cloning with that line in the gitattributes which continues to normalize line endings.

Makes sense, I'll try once more.

JunielKatarn commented:
> I still had the * text=auto in my repo when I cloned it, so not sure if this is a bug or if I'm not doing something I should be or what is happening heh

Can you try doing it directly on the Windows terminal (i.e. PowerShell)?
Maybe the issue only shows up on a fully native windows environment.

flagrama commented:
> I still had the * text=auto in my repo when I cloned it, so not sure if this is a bug or if I'm not doing something I should be or what is happening heh

Perhaps you have something set in your global git config? Curious what your `git config --list` presents.

JunielKatarn commented:
> > I still had the * text=auto in my repo when I cloned it, so not sure if this is a bug or if I'm not doing something I should be or what is happening heh
> 
> Perhaps you have something set in your global git config? Curious what your `git config --list` presents.

```
diff.astextplain.textconv=astextplain
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
http.sslbackend=openssl
http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
core.autocrlf=false
core.fscache=true
core.symlinks=true
pull.rebase=false
credential.helper=manager
user.name=JunielKatarn
core.autocrlf=false
```

vgturtle127 commented:
@JunielKatarn That's what I did, same result. It could be

```core.autocrlf=false```

JunielKatarn commented:
> @JunielKatarn That's what I did, same result. It could be
> 
> `core.autocrlf=false`

It's not. I just tried removing it before getting a clean clone.
I'll try cloning a remote branch that ONLY removes the text=auto entry.

flagrama commented:
> JunielKatarn That's what I did, same result. It could be
> 
> `core.autocrlf=false`

I think that might be accurate. I think that does the opposite of what * text does, and doesn't even normalize on commit.

flagrama commented:
> > @JunielKatarn That's what I did, same result. It could be
> > `core.autocrlf=false`
> 
> It's not. I just tried removing it before getting a clean clone.
> I'll try cloning a remote branch that ONLY removes the text=auto entry.

You have it set twice in your config. did you remove both of them?

JunielKatarn commented:
> > > @JunielKatarn That's what I did, same result. It could be
> > > `core.autocrlf=false`
> > 
> > 
> > It's not. I just tried removing it before getting a clean clone.
> > I'll try cloning a remote branch that ONLY removes the text=auto entry.
> 
> You have it set twice in your config. did you remove both of them?

Oh! It IS still there. I removed it from my user-level .gitconfig. I'll look where is the other one set and get back to you guys.

JunielKatarn commented:
> > > > @JunielKatarn That's what I did, same result. It could be
> > > > `core.autocrlf=false`
> > > 
> > > 
> > > It's not. I just tried removing it before getting a clean clone.
> > > I'll try cloning a remote branch that ONLY removes the text=auto entry.
> > 
> > 
> > You have it set twice in your config. did you remove both of them?
> 
> Oh! It IS still there. I removed it from my user-level .gitconfig. I'll look where is the other one set and get back to you guys.

I just tried using a Linux VM (ext4 FS, no special config).
Got this:
```
user.email=<my email>
user.name=JunielKatarn
```

.... same result. It's very puzzling that it only happens to me. I DID notice removing the .gitattributes entries for the affected file types also seems to fix it, but there may be a better way.

JunielKatarn commented:
Issue seems fixed by scoping the `text=auto` attribute to only a subset of the extensions.

What I tried for ultimate verification of the issue's existence:
1. Completely empty Linux environment (Live CD), installed git, cloned Pj64, no git status issues.
2. Completely empty Windows VM, installed Git, cloned Pj64, **the issue shows up in the terminal**.

Ready to merge with minimal changes.

vgturtle127 commented:
So this now just removes * text=auto?

JunielKatarn commented:
> So this now just removes * text=auto?

Not completely. It scopes it to specific extensions that require it (C++, VS props, installer, VS solution...).

Still a small change.

vgturtle127 commented:
Ah, so it only automatically normalizes certain file extensions. Gotcha. Alright, I guess that will work for now at least.

vgturtle127 commented:
You might consider renaming the PR title and such to reflect the changes

Jj0YzL5nvJ reviewed (changes requested):
It's extremely easy to misinterpret what `core.autocrlf` is actually doing, under normal conditions it's best left it unset. By default it only has an effect on new local repositories, not remote ones. Windows users see CRLF files even if they are LF and Not-Windows see LF files even if they are CRLF. Extensions at `.gitattributes` are the exceptions to the rule and are auto corrected if the encode is changed.

The problems start when you "correct a file encode" without a entry in `.gitattributes`. I don't see `.gitattributes` inside `.gitattributes`... Why are you correcting the encode of a file whose encode will be lost when it's cloned again? Is useless.

BTW, add:
```
*.cmd text eol=crlf
*.vbs text eol=crlf
*.pj.Lang text eol=crlf
```

JunielKatarn commented:
Just to be quadruple-sure about the issue:

- 100% fresh Windows install
- git version 2.30.0.windows.1
  - Configuration untouched

![image](https://user-images.githubusercontent.com/4507319/111038512-434c8a00-83de-11eb-9ded-d80cf13f76c5.png)

vgturtle127 commented:
.gitattributes can handle line endings but not charsets, but the EditorConfig can handle charsets as well.

As for what you were saying about the .gitattributes, we can add things for that as well pretty easily. But what @JunielKatarn seems to be going through doesn't seem to be consistent with what files are affected, unless I'm wrong about that.

Jj0YzL5nvJ commented:
- 999% old rancid Xubuntu 16.04.7 LTS install (4.15.0-136-generic / *hwe-16.04)
- git version 2.20.1 (own local build)
  - <details>
     <summary>Configuration untouched (nonexistent)</summary>

    <img src="https://0x0.st/-ZGe.png" alt="" class="">
    <img src="https://0x0.st/-ZG2.png" alt="" class="">
    <img src="https://0x0.st/-ZG_.png" alt="" class="">
    <img src="https://0x0.st/-Zda.png" alt="" class="">
    <img src="https://0x0.st/-ZdB.png" alt="" class="">

    </details>

 Erroneous encodings are now flagged based on the new rules. The next step is to reaffirm the rules by correcting the encode or correcting the rules for doing the wrong things.

vgturtle127 commented:
So you're saying that we need to add more things to the .gitattributes right? Then afterwards force the changes?

vgturtle127 commented:
Also we can add this to the EditorConfig @JunielKatarn :

```
[*.pj.Lang]
end_of_line = crlf
charset = utf-8-bom
```

Jj0YzL5nvJ commented:
I pointed out `*.cpp` files with CRLF encoding on Unix system, that's wrong. Therefore, it's necessary to correct these files.
Some others may be outside the scope of the rules, I don't know.

vgturtle127 commented:
Ohhhh, I see.

Yeah, I haven't been able to duplicate that issue and I'm also on Windows.

I never had the same issue as @JunielKatarn , so I can't comment much, but I haven't had any issues on my git cloned repos and branches with the ```* text=auto``` added

JunielKatarn commented:
> I pointed out `*.cpp` files with CRLF encoding on Unix system, that's wrong. Therefore, it's necessary to correct these files.
> Some others may be outside the scope of the rules, I don't know.
> 
> > I don't see .gitattributes inside .gitattributes...
> 
> Re-adding `* text=auto` can solve a LOT of things.

Why is that wrong?

It's been ages since any relevant editor such as VI or Visual Studio Code can handle those correctly (LF, CRLF, CR... your choice).
There is absolutely no need to have the line endings treated differently per platform (unless you want to default to LF and develop in, say, Windows 2000).

Jj0YzL5nvJ commented:
> Why is that wrong?

$ git config --global core.autocrlf true
$ git clone https://github.com/gonetz/GLideN64.git
```
Clonando en 'GLideN64'...
remote: Enumerating objects: 221, done.
remote: Counting objects: 100% (221/221), done.
remote: Compressing objects: 100% (96/96), done.
remote: Total 21682 (delta 144), reused 173 (delta 125), pack-reused 21461
Recibiendo objetos: 100% (21682/21682), 18.99 MiB | 2.24 MiB/s, listo.
Resolviendo deltas: 100% (16572/16572), listo.
```
$ cd GLideN64/projects/cmake/
$ cmake -DMUPENPLUSAPI=On ../../src/
```
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Deprecation Warning at GLideNHQ/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Found Freetype: /usr/lib/x86_64-linux-gnu/libfreetype.so  
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so   
Interprocedural optimizations enabled
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/build/GLideN64/projects/cmake
```
$ make -j5
```
Scanning dependencies of target osal
[  1%] Building C object osal/CMakeFiles/osal.dir/osal_files_unix.c.o
[  2%] Building C object osal/CMakeFiles/osal.dir/osal_keys_linux.c.o
[  3%] Linking C static library libosal.a
[  3%] Built target osal
Scanning dependencies of target GLideNHQ
[  4%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_hq2x.cpp.o
[  4%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters.cpp.o
[  5%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_hq4x.cpp.o
[  6%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_2xsai.cpp.o
[  7%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_xbrz.cpp.o
[  7%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxCache.cpp.o
[  8%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxDbg.cpp.o
[  9%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxFilter.cpp.o
[ 10%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxFilterExport.cpp.o
[ 11%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxHiResCache.cpp.o
[ 11%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxImage.cpp.o
[ 12%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxQuantize.cpp.o
[ 13%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxReSample.cpp.o
[ 14%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxTexCache.cpp.o
[ 15%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxUtil.cpp.o
[ 15%] Linking CXX static library libGLideNHQ.a
[ 15%] Built target GLideNHQ
[ 16%] Generate Git version
: not foundGLideN64/src/getRevision.sh: 4: /tmp/build/GLideN64/src/getRevision.sh: 
current revision "92c9bd84"
last build revision 
: not foundGLideN64/src/getRevision.sh: 7: /tmp/build/GLideN64/src/getRevision.sh: 
/tmp/build/GLideN64/src/getRevision.sh: 13: /tmp/build/GLideN64/src/getRevision.sh: Syntax error: end of file unexpected (expecting "then")
CMakeFiles/mupen64plus-video-GLideN64.dir/build.make:80: fallo en las instrucciones para el objetivo 'Revision.h'
make[2]: *** [Revision.h] Error 2
CMakeFiles/Makefile2:136: fallo en las instrucciones para el objetivo 'CMakeFiles/mupen64plus-video-GLideN64.dir/all'
make[1]: *** [CMakeFiles/mupen64plus-video-GLideN64.dir/all] Error 2
Makefile:102: fallo en las instrucciones para el objetivo 'all'
make: *** [all] Error 2
```

vgturtle127 commented:
Well, IIRC it's always been normal for files encoded on non-Windows systems to use LF and for Windows to use CRLF.

And most applications on either OS automatically change line endings for whatever reason.

IDK, this conversation is getting kind of over my head, I just wanted the .gitattributes file to automatically normalize line endings to make the transition to a cross-platform application easier and to make it easier for people on other platforms to contribute. 😕

JunielKatarn commented:
> > Why is that wrong?
> 
> $ git config --global core.autocrlf true
> $ git clone https://github.com/gonetz/GLideN64.git
> 
> ```
> Clonando en 'GLideN64'...
> remote: Enumerating objects: 221, done.
> remote: Counting objects: 100% (221/221), done.
> remote: Compressing objects: 100% (96/96), done.
> remote: Total 21682 (delta 144), reused 173 (delta 125), pack-reused 21461
> Recibiendo objetos: 100% (21682/21682), 18.99 MiB | 2.24 MiB/s, listo.
> Resolviendo deltas: 100% (16572/16572), listo.
> ```
> 
> $ cd GLideN64/projects/cmake/
> $ cmake -DMUPENPLUSAPI=On ../../src/
> 
> ```
> -- The C compiler identification is GNU 5.5.0
> -- The CXX compiler identification is GNU 5.5.0
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: /usr/bin/cc - skipped
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Check for working CXX compiler: /usr/bin/c++ - skipped
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> CMake Deprecation Warning at GLideNHQ/CMakeLists.txt:1 (cmake_minimum_required):
>   Compatibility with CMake < 2.8.12 will be removed from a future version of
>   CMake.
> 
>   Update the VERSION argument <min> value or use a ...<max> suffix to tell
>   CMake that the project does not need compatibility with older versions.
> 
> 
> -- Found Freetype: /usr/lib/x86_64-linux-gnu/libfreetype.so  
> -- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so   
> Interprocedural optimizations enabled
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /tmp/build/GLideN64/projects/cmake
> ```
> 
> $ make -j5
> 
> ```
> Scanning dependencies of target osal
> [  1%] Building C object osal/CMakeFiles/osal.dir/osal_files_unix.c.o
> [  2%] Building C object osal/CMakeFiles/osal.dir/osal_keys_linux.c.o
> [  3%] Linking C static library libosal.a
> [  3%] Built target osal
> Scanning dependencies of target GLideNHQ
> [  4%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_hq2x.cpp.o
> [  4%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters.cpp.o
> [  5%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_hq4x.cpp.o
> [  6%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_2xsai.cpp.o
> [  7%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TextureFilters_xbrz.cpp.o
> [  7%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxCache.cpp.o
> [  8%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxDbg.cpp.o
> [  9%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxFilter.cpp.o
> [ 10%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxFilterExport.cpp.o
> [ 11%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxHiResCache.cpp.o
> [ 11%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxImage.cpp.o
> [ 12%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxQuantize.cpp.o
> [ 13%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxReSample.cpp.o
> [ 14%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxTexCache.cpp.o
> [ 15%] Building CXX object GLideNHQ/CMakeFiles/GLideNHQ.dir/TxUtil.cpp.o
> [ 15%] Linking CXX static library libGLideNHQ.a
> [ 15%] Built target GLideNHQ
> [ 16%] Generate Git version
> : not foundGLideN64/src/getRevision.sh: 4: /tmp/build/GLideN64/src/getRevision.sh: 
> current revision "92c9bd84"
> last build revision 
> : not foundGLideN64/src/getRevision.sh: 7: /tmp/build/GLideN64/src/getRevision.sh: 
> /tmp/build/GLideN64/src/getRevision.sh: 13: /tmp/build/GLideN64/src/getRevision.sh: Syntax error: end of file unexpected (expecting "then")
> CMakeFiles/mupen64plus-video-GLideN64.dir/build.make:80: fallo en las instrucciones para el objetivo 'Revision.h'
> make[2]: *** [Revision.h] Error 2
> CMakeFiles/Makefile2:136: fallo en las instrucciones para el objetivo 'CMakeFiles/mupen64plus-video-GLideN64.dir/all'
> make[1]: *** [CMakeFiles/mupen64plus-video-GLideN64.dir/all] Error 2
> Makefile:102: fallo en las instrucciones para el objetivo 'all'
> make: *** [all] Error 2
> ```

I see your point. `core.autocrlf` just shouldn't be true by default. Not anymore.
It's used based on the utterly false assumption that Windows can't handle "UNIX" line endings or viceversa.

That said, it IS the default, so I'll gladly apply whatever makes sense to most devs.
The bottom line is, "unfixable" checked out files on a clean clone need to go.

Jj0YzL5nvJ commented:
Interpreted code (glue code) needs the correct EOL. Well, in Windows `core.autocrlf` the default is true but in Unix is false.

JunielKatarn commented:
> Well, IIRC it's always been normal for files encoded on non-Windows systems to use LF and for Windows to use CRLF.
> 
> And most applications on either OS automatically change line endings for whatever reason.
> 
> IDK, this conversation is getting kind of over my head, I just wanted the .gitattributes file to automatically normalize line endings to make the transition to a cross-platform application easier and to make it easier for people on other platforms to contribute. 😕

Maybe it's best to take this in depth to Discord.

I have a more pragmatic perspective about this change.
Whatever people (mostly @project64) prefer that will fix the "tainted checkout" issue, I'll go for.

Do note, setting core.autocrlf to either true or false should not taint cloned repos. If it does, the it's the repo config that is wrong.

Jj0YzL5nvJ commented:
I would be more than satisfied with making Project64 compatible with Wine =P

vgturtle127 commented:
@JunielKatarn Me and Zilmar are discussing it a bit now actually. He wants to take a look at it in the next day or so if we don't work it out.

He said that he had a build fail with the line endings being the cause of the issue, so perhaps we should wait and see what Zilmar can find out, since he can at least somewhat replicate a similar issue, while I can't on my end.

And to be honest, when I changed this I did know it might cause some issues, I just didn't know it would be something like this. Often times after you introduce automatic line ending normalization it just tells you that files are changed (and sometimes they are) and you just commit those changes and then sort of forget about it, but apparently Zilmar is experiencing something where builds fail, so I'm sort of stuck right now.

I will continue to work on some of my busy work in the mean time, and I think I'll wait for Zilmar to weigh in on it and see what he can find. 😄

At the end of the day my goal with the change was so that it would be easier to contribute on systems where Project64 has potential to be ported, like Linux, Android, and possibly even macOS. And of course working on 64-bit when possible.

If the change causes too many issues, I'm OK with reverting some things until we can figure out a better solution.

I may take a look at other projects and see what they do in regards to this, although most of them start from the ground up with cross-platform development in mind and they usually have developers on Windows and Linux, so perhaps that is something to consider.

flagrama commented:
> I would be more than satisfied with making Project64 compatible with Wine =P

It already is. It's how I usually use Project 64. I use it via Steam's Proton specifically, but it runs fine.

flagrama commented:
> > I pointed out `*.cpp` files with CRLF encoding on Unix system, that's wrong. Therefore, it's necessary to correct these files.
> > Some others may be outside the scope of the rules, I don't know.
> > > I don't see .gitattributes inside .gitattributes...
> > 
> > 
> > Re-adding `* text=auto` can solve a LOT of things.
> 
> Why is that wrong?
> 
> It's been ages since any relevant editor such as VI or Visual Studio Code can handle those correctly (LF, CRLF, CR... your choice).
> There is absolutely no need to have the line endings treated differently per platform (unless you want to default to LF and develop in, say, Windows 2000).

The editors can read either line ending. They still write the platform-specific line ending. Getting mixed line endings eventually leads to annoying to read diffs.

Jj0YzL5nvJ commented:
> It already is. It's how I usually use Project 64. I use it via Steam's Proton specifically, but it runs fine.

Does it work in 6.X? I need to upgrade then, I got stuck at 5.0.3.

flagrama commented:
I use some version of 5.0. I've had issues with newer versions. You're running the 2.4 dev builds of Project 64 and not 2.3, right?

Jj0YzL5nvJ commented:
It always crashes, hangs or freezes...

With GLideN64
```
002b:fixme:ver:GetCurrentPackageId (0x18aff0c (nil)): stub
0009:fixme:resource:DestroyAcceleratorTable other process handle 0x1?
0009:fixme:sync:NtSetTimerResolution (10000,1,0x31bf88), stub!
0009:fixme:win:RegisterDeviceNotificationW (hwnd=0xa008c, filter=0x31ba44,flags=0x00000004) returns a fake device notification handle!
0032:fixme:ver:GetCurrentPackageId (0x19eff0c (nil)): stub
Could not open font C:\windows/Fonts/arial.ttf
Terminado (killed)
```

With Jabo
```
002b:fixme:ver:GetCurrentPackageId (0x127ff0c (nil)): stub
0009:fixme:resource:DestroyAcceleratorTable other process handle 0x1?
0009:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
0009:fixme:d3d:wined3d_adapter_init_gl_caps A set of 3256072 devices is not supported.
0009:fixme:d3d:wined3d_adapter_init_gl_caps A set of 3258376 devices is not supported.
0009:fixme:sync:NtSetTimerResolution (10000,1,0x31c010), stub!
0009:fixme:win:RegisterDeviceNotificationW (hwnd=0x6007e, filter=0x31bac4,flags=0x00000004) returns a fake device notification handle!
Terminado (killed)
```

With "Project64 Video" http://0x0.st/-ZnP.png
```
002c:fixme:ver:GetCurrentPackageId (0xe1ff0c (nil)): stub
0009:fixme:resource:DestroyAcceleratorTable other process handle 0x1?
0009:fixme:sync:NtSetTimerResolution (10000,1,0x31c010), stub!
0009:fixme:win:RegisterDeviceNotificationW (hwnd=0x50086, filter=0x31bac4,flags=0x00000004) returns a fake device notification handle!
wine: Unhandled exception 0xc0000417 in thread 9 at address 00530B89 (thread 0009), starting debugger...
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
0032:fixme:dbghelp:elf_search_auxv can't find symbol in module
```

flagrama commented:
Yeah, I dunno. Works for me.
![image](https://user-images.githubusercontent.com/634079/111059242-f7dbbf80-8461-11eb-9016-ae2fc79e20a9.png)

You have all the vulkan drivers you need installed, right?

Jj0YzL5nvJ commented:
That's Proton, not Wine ¬¬
I prefer not to install a lot of things in the user space...

> You have all the vulkan drivers you need installed, right?

Yes... I can't run ParaLLEl-RDP though (I don't have support for the VK_KHR_8bit_storage extension)

JunielKatarn commented:
> The editors can read either line ending. They still write the platform-specific line ending. Getting mixed line endings eventually leads to annoying to read diffs.

Is this accurate?
Editors (again, in any non-ancient OS) can read AND write any line ending.
Moreover, the EditorConfig-aware ones will just do whatever the config tells them automatically.

What editors that a significant amount of Project64 contributors would be using is unable to emit specified line endings or be compatible with EditorConfig?

Jj0YzL5nvJ commented:
From my perspective, the default settings for `core.autocrlf` and `.gitattributes` have a very simple mission. **Anti-PEBCAK**.
If you know absolutely nothing about EOL and what `git` really is doing behind the scene within each OS, you are truly happy (ignorance is bliss). Everything works because `git` is holding a dream in which you can't see reality.

Messing with `core.autocrlf` and `.gitattributes` means that you know reality and are forcing others to see it partially or totally. The ideal is to create a balance in which both can coexist. Users who know about EOL and those who don't.

If me, a Unix user who knows about EOL problems. I want to contribute to a project that does not use `.gitattributes` with something intended for Windows users. [The best thing I can do before contributing is to use `git config --global core.autocrlf true`.](https://github.com/gonetz/GLideN64/blob/master/projects/msvc/GLideN64_builder.cmd)

project64 commented:
I am just being cautious with merging any project settings before I finalize 3.0, which is why I am closing with out merging

JunielKatarn commented:
> I am just being cautious with merging any project settings before I finalize 3.0, which is why I am closing with out merging

@project64 this change fixes a very recent commit that introduces a repository working directory state issue.
This change is necessary to allow people to check out files and avoid committing problems.
Can you reconsider this one? It does not affect the compilation or linking process in any way. It's strictly related to source code management.

@vgturtle127 can you perhaps vouch for this change?

project64 commented:
this might be correct, I have removed .gitattributes for now

vgturtle127 commented:
Yeah, we might look into adding it back after 3.0 is released
